hoomd/simulation.py

```diff
@@ -32,6 +32,7 @@ def __init__(self, device):
         self._operations = Operations()
         self._operations._simulation = self
         self._timestep = None
+        self._seed = 0
```

Comment:
Does it make sense to default the seed to zero or to a random seed chosen from the system entropy? The former is more explicit for the user, but the latter is more like what is done by numpy. (i.e., don't set the seed unless you care about reproducing the simulation.)

Reply:
My preference is that "explicit is better than implicit" and script runs should be as reproducible as possible.  In hoomd v2, I required the seed argument to every constructor. I don't want to do that now because some users may run simulations that don't use random numbers and being forced to provide a seed would seem pointless to them. 

However, if others in the dev community prefer system entropy selected seeds I can be outvoted. What does everyone think? @glotzerlab/hoomd-api-reviewers

I should at least document the default seed behavior.

Reply:
I am OK either way. I can imagine the post, "Help! My Brownian dynamics simulation is always giving the same answer!" popping up quite a bit :-), but I myself always use a fixed seed when I am debugging and writing my production scripts. For playing around, though, it is nice to have numpy generate a new random stream automatically.

On the whole, I think I would vote to randomize the seed and let the user override this if they want to set it for reproducibility, but I'm OK with whatever gets decided.

Reply:
> I think I would vote to randomize the seed and let the user override this if they want to set it for reproducibility

Agreed. This is what I would expect by default.

Reply:
I talked with a few other developers offline. Many of them like the additional reproducibility that an explicit seed provides. Users who wish to use system entropy can opt in with `seed=secrets.randbits(16)`.

Reply:
Is a simulation reproducible with the same seed and different devices? (My understanding is that the answer is “no.”) I have concerns that the implications of perfect-reproducibility-by-default-seed may be too strong.

Reply:
If this is the goal, it seems like it would be better to force the user to set the seed (reproducible and explicit) rather than take some default value (reproducible but implicit). I agree that it's confusing to require a seed in simulations that don't consume random numbers, though, so this would require registering classes that use random numbers and checking that the seed has been properly set.

Reply:
@bdice These are always reproducible across versions and across compilers and devices:
* Randomized velocities
* Randomized thermostats
* Random forces and trial moves in the first time step

Due to their chaotic behavior, simulation trajectories will diverge due to floating point errors after a few hundred or thousand steps (HPMC trajectories will be stable for longer than MD). Different compilers and/or different devices will lead to different orders for floating point operations that will cause trajectory divergence. However, for multiple runs with the same build of HOOMD (e.g. container) on the same device (CPU or GPU), the following will be binary reproducible:

* HPMC trajectories on the CPU and GPU
* MD trajectories on the CPU
* MD trajectories on the GPU (but only when the user opts into a slower deterministic mode).

@mphoward That is an interesting idea and may be doable. I wonder how users with working scripts will react when they add a new operation or method that requires a seed and then get an error. Is that better or worse than requiring users to set a seed even when they know it won't be used? 

We could make it a warning instead of an error. Then the simulation would run with a default seed and notify the user that this is the case.

Reply:
Great info, @joaander, that helps. I really like your idea of warning the user if the seed is left at the default value. That seems like the best tradeoff of user responsibility and user expectations.